MQTT patches

mike-scott opened the pull request:
2 of these 3 patches are needed to use MQTT with the FOTA app changes that we have in our zephyr-fota-hawkbit repository.  The 1 patch not needed is part of a clean up series that is WIP, but won't break anything here.

Commit:
net/mqtt: combine mqtt_parser for PUBLISHER and SUBSCRIBER

Currently, app type MQTT_APP_PUBLISHER_SUBSCRIBER is not supported.
mqtt_init() will return EINVAL because the context rcv function is
either mqtt_publisher_parser or mqtt_subscriber_parser (not both).

Let's combine these functions into mqtt_parser and remove the app
type check from mqtt_init().

Change-Id: I60460e011395864706e293c997e9f8a65681b368
Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
net/mqtt: simplify parser packet specific function calls

Let's remove the 4 publish rx wrapper functions which are only called
in the combined mqtt_parser() function and replace them with
mqtt_rx_pub_msgs().

Change-Id: I7b11e930f8976d16380f6b1c2c85537211bcde3a
Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
net/mqtt: use rlen to calculate # of QoS items in SUBACK packet

In testing, the data->len of a netbuf can be slightly larger than
the actual MQTT packet for some reason.  Example we've seen in
is that 2 SUBACK packets are both inbound and the first has a
data->len of 5 and the second has a data->len of 7.

Under this case we end up with the second packet looking for 3
QoS items when there is only 1.

We should not be using the data->len to calculate the # of QoS
items in the SUBACK packet.  Instead, let's use the remaining length
parsed from the end of the SUBACK packet.  This value minus the
packet id length (2 in the current spec) should be the # of QoS items
left to read in the packet.

Change-Id: I7f6cebaaed9570b778d466de84331cf8c5060755
Signed-off-by: Michael Scott <michael.scott@linaro.org>

gatek33p3r commented:
Checkpatch Suceeded.

gatek33p3r commented:
Sanity Check Failed.

mike-scott commented:
Woops, this failed because I had some debugging left in 1 of the patches which needs to be removed.  I'll repost.

gatek33p3r commented:
Checkpatch Suceeded.

gatek33p3r commented:
Sanity Check Succeeded.

mbolivar commented on subsys/net/lib/mqtt/mqtt.c:
```diff
@@ -843,17 +786,7 @@ int mqtt_init(struct mqtt_ctx *ctx, enum mqtt_app app_type)
-	default:
-		return -EINVAL;
-	}
+	ctx->rcv = mqtt_parser;
```
There's a race here: rcv should be installed before mqtt_recv is passed to net_context_recv().

mike-scott replied:
Yep, you're right.  I should probably put in a separate patch to fix that or add that fix first and then lay this patch on top.  If you want I can stick you in as author since it would be a code shuffle only patch (easy).

mbolivar replied:
Ah, for some reason I didn't see that this was there before you did anything -- no need to make me an author since you did the work. Thanks though!

mbolivar commented on subsys/net/lib/mqtt/mqtt.c:
```diff
@@ -726,70 +727,12 @@ int mqtt_publisher_parser(struct mqtt_ctx *ctx, struct net_buf *rx)
-	case MQTT_PINGRESP:
-		rc = mqtt_rx_pubrel(ctx, data);
-		break;
 	case MQTT_SUBACK:
```
Combining these into one function seems like it turns every app into an MQTT_APP_PUBLISHER_SUBSCRIBER. A publish-only app wouldn't expect to ever receive SUBACK, so I'd expect this to error out here in that case, and similarly in other situations (receiving PUBACK or PUBCOMP in a subscribe-only app, etc.)

Am I missing something?

mbolivar replied:
(Not sure why GitHub is showing this as "outdated"...)

mbolivar commented on subsys/net/lib/mqtt/mqtt_pkt.c:
```diff
@@ -815,7 +815,7 @@ int mqtt_unpack_suback(uint8_t *buf, uint16_t length, uint16_t *pkt_id,
 	offset += PACKET_ID_SIZE;
 
-	*items = length - offset;
+	*items = rlen - PACKET_ID_SIZE;
```
This line looks good, but wanted to make a note about the commit message: what I remember seeing was a SUBACK and a PINGRESP combined in a single net_buf (IIRC they were in the same IPv6 packet). So first SUBACK (SUBACK variable header + one QoS item), the next MQTT packet had length 2 (0xd000 == PINGRESP), leading to 3 items (for the one QoS item and the two bytes in the PINGRESP). Did you see the two-SUBACKS-in-a-single-packet in your own testing?

mike-scott replied:
Hah nice catch.   I wrote the patch after some initial debugging (a different way) and then adjusted it later once we realized the actual cause.  Patch message didn't get that change.

Commit:
net/mqtt: remove MQTT app_type

Now that SUBSCRIBER and PUBLISHER functions have been combined, we can
remove the app_type and associated checks.

Later we will add packet_id management which will reject *ACK messages
for packets not originating from the Zephyr app.  This should clear up
any older checks which rejected messages based on app type.

Change-Id: I221a013c06f7890f0ba4c9a0880b06f8356214dc
Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
net/mqtt: fix race condition in mqtt_init()

We need to set the mqtt context recv function prior to calling
net_context_recv which installs the mqtt_recv callback.
If not, we risk a race condition where an unprotected reference
to mqtt->rcv(mqtt, buf) is made in mqtt_recv().

Change-Id: If90ee58f4ea6f7879ef7c12b969ba27647426acc
Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

gatek33p3r commented:
Checkpatch Suceeded.

gatek33p3r commented:
Sanity Check Succeeded.

mbolivar commented:
LGTM.

mike-scott commented:
Reposting
